Use cloud-hosted collections only

chuckwondo opened the pull request:
All supported GEDI collections are now cloud-hosted, which means we can pull granules via S3.

This PR includes the following:

- updates the CMR queries to obtain the cloud-hosted collections only (#65)
- adds vcrpy for unit testing CMR queries
- adds a GitHub Actions workflow to automatically run unit tests and linting on PRs (#24)
- removes a few more references to the `returns` library (#3)

wildintellect commented on bin/create:
What happens if `libmamba` isn't installed?

chuckwondo replied:
The `libmamba` solver is now the default solver with `conda` (https://conda.org/blog/2023-11-06-conda-23-10-0-release/), and:

> Since July 2023, the [conda-libmamba-solver](https://github.com/conda/conda-libmamba-solver) plugin has been included in all major conda ecosystem installers ([miniforge](https://conda-forge.org/miniforge/), [miniconda](https://docs.conda.io/projects/miniconda/en/latest/index.html), [mambaforge](https://conda-forge.org/miniforge/) and [Anaconda Distribution](https://www.anaconda.com/download)), but was disabled by default. As soon as these installers are updated to contain conda 23.10.0 or later, they will automatically default to using the conda-libmamba-solver plugin.

wildintellect commented on bin/run:
```diff
@@ -28,4 +28,9 @@ if ! "${conda}" env list | grep -q gedi_subset; then
+if [[ "${CI:-}" == "true" ]]; then
+    "${conda_run_args[@]}"
+else
+    "${conda}" run --live-stream --name gedi_subset "${conda_run_args[@]}"
```
looks like `--live-stream` is being deprecated https://github.com/conda/conda/issues/11209
change to `--no-capture-output` ? not sure if that has the same intent.

chuckwondo replied:
Too funny. A while back, I was using `--no-capture-output`. I don't recall why I landed on using `--live-stream` instead, but I'll switch back :-)

wildintellect commented on CONTRIBUTING.md:
Should you mention the use of VCR in the tests, since this avoids the need to actually query CMR?

chuckwondo replied:
Not quite sure how to include a blurb about this without getting into too much detail.

Also, querying the CMR is _initially_ needed when adding new tests because vcrpy must perform a real query, which it then records. Subsequent runs of tests that have already been recorded do not requery.

This means that when adding _new_ tests that query CMR, they must initially be run within the ADE. Note, however, once we make use of say `earthaccess` instead of `maapy-py` for CMR queries, it should no longer be necessary to run within the ADE.

I'll add a brief blurb.

wildintellect commented:
I got a lint error, and 1 warning. Should we fix before accepting?
Linter caught in /tests/conftest.py:
```
             self._MAAP_HOST = "api.maap-project.org"
-            self._SEARCH_COLLECTION_URL = f"https://{self._MAAP_HOST}/api/cmr/collections"
+            self._SEARCH_COLLECTION_URL = (
+                f"https://{self._MAAP_HOST}/api/cmr/collections"
+            )
             self._CMR = CMR([], 20, {})
             self.aws = AWS(
```         
             
Warnings         
 ```
tests/test_gedi_utils.py::test_subset_hdf5[lat_lowestmode-lon_lowestmode-all-columns0-sensitivity < 0.9-0]
  /home/ochotona/code/devseed/maap/gedi-subsetter/tests/test_gedi_utils.py:222: FutureWarning: <class 'geopandas.array.GeometryArray'>._reduce will require a `keepdims` parameter in the future
    assert gdf.notna().all(axis=None)
```

wildintellect commented on CHANGELOG.md:
should anything about the github actions and changes to tests be mentioned in the Changelog?

chuckwondo commented:
> > I got a lint error, and 1 warning. Should we fix before accepting? Linter caught in /tests/conftest.py:
> > ```
> >              self._MAAP_HOST = "api.maap-project.org"
> > -            self._SEARCH_COLLECTION_URL = f"https://{self._MAAP_HOST}/api/cmr/collections"
> > +            self._SEARCH_COLLECTION_URL = (
> > +                f"https://{self._MAAP_HOST}/api/cmr/collections"
> > +            )
> >              self._CMR = CMR([], 20, {})
> >              self.aws = AWS(
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Warnings
> > ```
> > tests/test_gedi_utils.py::test_subset_hdf5[lat_lowestmode-lon_lowestmode-all-columns0-sensitivity < 0.9-0]
> >  /home/ochotona/code/devseed/maap/gedi-subsetter/tests/test_gedi_utils.py:222: FutureWarning: <class 'geopandas.array.GeometryArray'>._reduce will require a `keepdims` parameter in the future
> >    assert gdf.notna().all(axis=None)
> > ```
> 
> Ah, I missed that formatting error because that was a change I pushed from the ADE, where pre-commit didn't catch that. I'll fix it.

Actually, the build should have failed on that because pre-commit.ci should have been triggered, but it wasn't. Let me see if I can get pre-commit.ci to run as expected.

chuckwondo commented:
@wildintellect, I added a blurb on vcrpy and I got pre-commit.ci to trigger. It turned out that there's a setup step that is not clearly indicated on the pre-commit.com site. Essentially, each repo must give permission to pre-commit.ci, in _addition_ to adding a `ci` block to `.pre-commit-config.yaml`.

Gah! Looks like pre-commit failed. At least it was triggered! Let me see what happened.

chuckwondo commented:
> @wildintellect, I added a blurb on vcrpy and I got pre-commit.ci to trigger. It turned out that there's a setup step that is not clearly indicated on the pre-commit.com site. Essentially, each repo must give permission to pre-commit.ci, in _addition_ to adding a `ci` block to `.pre-commit-config.yaml`.
> 
> Gah! Looks like pre-commit failed. At least it was triggered! Let me see what happened.

Oooh! I see what happened.  I have to tweak the pre-commit file. I had added a hook that ends up calling `conda`, which isn't installed on the pre-commit.ci server. Let me address that.

chuckwondo commented:
@wildintellect, I had to wrestle a bit with pre-commit stuff since your earlier approval. Mind having another look?
